src/clojure/nrepl/middleware/session.clj

```diff
@@ -137,14 +172,51 @@
          (atom (merge baseline-bindings (dissoc (get-thread-bindings) #'*agent*))
                :meta {:id id
                       :stdin-reader stdin-reader
-                      :input-queue input-queue}))))))
+                      :input-queue input-queue
+                      :exec default-exec}))))))
+
+(defn session-exec [id]
+  (let [cl (clojure.lang.DynamicClassLoader.
+            (.getContextClassLoader (Thread/currentThread)))
+        queue (LinkedBlockingQueue.)
+        running (atom nil)
+        thread (atom nil)
+        main-loop #(let [[id ^Runnable r ^Runnable ack] (.take queue)]
+                     (reset! running id)
+                     (when (try
+                             (.run r)
+                             (compare-and-set! running id nil)
+                             (finally
+                               (compare-and-set! running id nil)))
+                       (some-> ack .run)
+                       (recur)))
+        spawn-thread #(doto (Thread. main-loop (str "nRepl-session-" id))
+                        (.setDaemon true)
+                        (.setContextClassLoader cl)
+                        .start)]
+    (reset! thread (spawn-thread))
+    {:interrupt (fn [exec-id] ; nil means interrupt whatever is running
+                  ; returns :idle, interrupted id or nil
+                  (let [current @running]
+                    (cond
+                      (nil? current) :idle
+                      (and (or (nil? exec-id) (= current exec-id)) ; cas only checks identity, so check equality first 
+                           (compare-and-set! running current nil))
+                      (do
+                        (doto ^Thread @thread .interrupt .stop)
+                        (reset! thread (spawn-thread))
+                        current))))
+     :close #(.interrupt ^Thread @thread)
+     :exec (fn [exec-id r ack]
+             (.put queue [exec-id r ack]))}))
 
 (defn- register-session
   "Registers a new session containing the baseline bindings contained in the
    given message's :session."
   [{:keys [session transport] :as msg}]
   (let [session (create-session transport @session)
-        id (-> session meta :id)]
+        {:keys [id]} (meta session)]
+    (alter-meta! session into (session-exec id))
```

Comment:
I wonder if we shouldn't tackle this in `create-session`, but it's not a big deal.

Reply:
How so? In `create-session` we don't know yet if the session is going to be long-lived (what's the current practice for client implementors? 1 long-lived session for all interactions?) that's why `create-session` defaults to threadpool execution.

Oh! Maybe you meant to pass a flag to `create-session`, is this it?

Reply:
Is `clone` ever used to fork a new long-lived session from another long-lived session?

Reply:
The option certainly exists. Don't know how many people make use of it. I know some people were complaining that cloning a long-lived session didn't copy the dynamic vars from the original session (which has been fixed since then), so I guess someone does this.